viewing_party/party.py

```diff
@@ -1,23 +1,320 @@
+
+    Returns:
+        list of dict of unique movies.
+    """
+    
+    user_watched = user_data["watched"]
+    friends_watched = get_friends_watched(user_data)
+
+    unique_friends_watched = []
+
+    for movie in friends_watched:
+        if movie not in user_watched and movie not in unique_friends_watched:
+            unique_friends_watched.append(movie)
+    
+    return unique_friends_watched
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+
+def get_available_recs (user_data):
+    """
+    Return a list of recommended movies based on the user's subscriptions 
+        and their's friends watched movies
+        
+    Parameters:
+        user_data (dict): has two keys:
+            "subscriptions": List of streaming services the user has access to
+            "friends": List of friends, each having a "watched" list of movie dictionaries, 
+                each movie has a "host"
+    
+    Returns:
+        list of dict of movies
+    """
+    
+    user_watched = user_data["watched"]
+    friends_watched = get_friends_watched(user_data)
+  
+    recommended_movies = []
+    for movie in friends_watched:
+        if movie not in user_watched and movie["host"] in user_data["subscriptions"] \
+            and movie not in recommended_movies:
+            recommended_movies.append(movie)
+    
+    return recommended_movies   
+           
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    """
+    Determine a list of recommended movies
+    
+    Parameters:
+        user_data (dict)
+    Notes:
+        user has not watched that movie
+        at leats one of user's friends has watched it
+        "genre" == user's frequent genre
+    Returns:
+        recomemded movies (list)
+    """
+    
+    user_watched = user_data["watched"]
+    friends_watched = get_friends_watched(user_data)
+    most_frequenty_genre = get_most_watched_genre(user_data)
+    
+    # Returns a list of recommended movies based in user's most frequently watched genre.
+    recommended_movies = []
+    for movie in friends_watched:
+        if movie not in user_watched and \
+            movie["genre"] == most_frequenty_genre and\
+            movie not in recommended_movies:
+
+            recommended_movies.append(movie)
+    
+    return recommended_movies  
+
+
+def get_rec_from_favorites(user_data):
+    """
+    Determine a list of recommended movies
+    
+    Parameters:
+        user_data (dict): "favorites" field contain a favorite list of movie dictionaries of user
+    Notes:
+        movie in user's "favorites"
+        non of user's friends have watched it
+        
+    Returns:
+        recomemded movies (list)
+    """
+    user_watched = user_data["watched"]

```

Comment:
We only use `user_watched` one other time in this function! When we use a value (`user_data["watched"]')  just a few times, it can be a good idea to just avoid using a variable as a middle man. Direct access will work just fine!